test/test_scripts_run_task.py

```diff
@@ -270,3 +272,56 @@ def _Popen(
 
     assert os.path.isdir(untracked_dir.name) is False
     assert os.path.isfile(untracked_file.name) is False
+
+
+@pytest.fixture(scope="session")  # Tests shouldn't change this repo
+def mock_git_repo():
+    "Mock repository with files, commits and branches for using as source"
+    with tempfile.TemporaryDirectory() as repo:
+        repo_path = str(repo)
+        subprocess.check_call(["git", "init", "-b", "main"], cwd=repo_path)
+
+        def _commit_file(message):
+            untracked_file = tempfile.NamedTemporaryFile(dir=repo, delete=False)
+            untracked_file.write(b"test file content")
+            untracked_file.close()
+            subprocess.check_call(["git", "add", "."], cwd=repo_path)
+            subprocess.check_call(["git", "commit", "-m", message], cwd=repo_path)
+
+        _commit_file("Initial commit")
+        subprocess.check_call(["git", "checkout", "-b", "mybranch"], cwd=repo_path)
+        _commit_file("File in my branch")
+        subprocess.check_call(["git", "checkout", "main"], cwd=repo_path)
+        yield repo_path
+
+
+@pytest.mark.parametrize(
+    "base_ref,ref",
```

Comment:
Should we also test passing a non-None commit?

Reply:
I understand this is going to be trickier to test defined commit hashes. On the bright side, commit hashes are predictable. We can set all parameters to known values and hardcode the hash we want to pass:

> The git hash is made up of the following:
>
> *    The commit message
> *    The file changes
> *    The commit author (and committer- they can be different)
> *    The date
> *    The parent commit hash

_Source: https://www.mikestreety.co.uk/blog/the-git-commit-hash/_

The parent commit hash is zero[1] when you create the first commit hash. `freezegun`[2] will help to set the date. The remaining params are already set by your test. 

With this, we should be able to test passing non-None commits that reflect how this function is called in real life 🙂 

[1] https://github.com/git/git/blob/dc04167d378fb29d30e1647ff6ff51dd182bc9a3/t/oid-info/hash-info#L7
[2] https://pypi.org/project/freezegun/

Reply:
I feel like this is over-engineering. For now I just pulled the hashes created by the fixture, we should be able to trust the `git` program, and the target of the test is to cover the business logic of run-task.

Maybe for a future release we could use a python library to handle git? https://gitpython.readthedocs.io/en/stable/tutorial.html and the testing wrapper https://pypi.org/project/pytest-git/

Reply:
> For now I just pulled the hashes created by the fixture, we should be able to trust the git program, and the target of the test is to cover the business logic of run-task.

I 100% agree with you 🙂 We shouldn't test `git` itself and we should limit ourselves to what `run-task` does. That said, this line: https://github.com/taskcluster/taskgraph/commit/df1b39b70b93a119e0d8dd64c3ceff180331836e#diff-e29e172140c458b95d3f5b1a5542188ffe334182ccef1c3b62e1eeb102beec1aR671 is part of `run-task` and we may regress it one day. So having an end-to-end test that ensures we are able to checkout a known revision will help us from regressing our business logic. 

In any case, I think you were right to merge this PR, yesterday! Without your PR, we didn't have any coverage there so it's definitely an improvement 👍 We can address the test in a follow-up but it can be done much later since we've lived without tests for year there. 

> Maybe for a future release we could use a python library to handle git?

Good idea! I think we should tackle this once we drop the hg support. The current logic between `hg` and `git` is very close. If we change `git` but not `hg`, it'll be harder to see patterns between the 2. 

How does that sound?

Reply:
Sounds good. Another way of testing this is mocking out the subprocess calls and check for units. Ie: Given input x,y,z we expect git to be called with commit/branch y. That way we don't have to fiddle with the git internals, we just "trust" it will work.

Reply:
Works for me too 🙂